test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+        // I have checked that in a normal signal handler I get the correct account pointer.
+        // auto accountFilesChanged = filesChangedSpy.at(0).at(0).value<OCC::Account *>();
+        // QCOMPARE(accountFilesChanged, account);
+    }
+
+    void testOnWebSocketTextMessageReceived_invalidCredentialsMessage_reconnectWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy spy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(spy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication attempt and then sent invalid credentials
+        spy.wait();
+        QCOMPARE(spy.count(), 2);
+        const auto socket = spy.at(0).at(0).value<QWebSocket *>();
+        const auto firstPasswordSent = spy.at(1).at(1).toString();
+        QCOMPARE(firstPasswordSent, password);
+        socket->sendTextMessage("err: Invalid credentials");
+
+        // Wait for a new authentication attempt
+        spy.wait();
+        QCOMPARE(spy.count(), 4);
+        const auto secondPasswordSent = spy.at(3).at(1).toString();
+        QCOMPARE(secondPasswordSent, password);
+    }
+
+    void testOnWebSocketError_connectionLost_emitConnectionLost()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+        QSignalSpy connectionLostSpy(pushNotifications.get(), &OCC::PushNotifications::connectionLost);
+        QVERIFY(connectionLostSpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication and then sent a network error
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 2);
+        // FIXME: This a little bit ugly but I had no better idea how to trigger a error on the websocket client.
+        // The websocket that is retrived through the server is not connected to the error signal.
+        emit pushNotifications->_webSocket->error(QAbstractSocket::SocketError::NetworkError);
+
+        // Wait for connectionLost signal
+        // FIXME: For some reason it takes a few seconds until this signal arrives. Why is this?
```

Comment:
Well I guess this is because you force the signal emit but the internal state of the socket didn't change one bit when you do so. This is a very synthetic test.